webhook/resourcesemantics/validation/validation_admit.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"go.uber.org/zap"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/apis"
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/webhook"
+	"knative.dev/pkg/webhook/resourcesemantics"
+)
+
+var errMissingNewObject = errors.New("the new object may not be nil")
+
+// Callback is a generic function to be called by a consumer of validation
+type Callback struct {
+	Func func(ctx context.Context, unstructured *unstructured.Unstructured) error
+
+	SupportedVerbs map[webhook.Operation]bool
```

Comment:
> I think keeping the map is desirable (really I want a Set) so you can't repeat verbs and makes it easier to not have to search the slice.

I still think the map is odd for folks constructing these callbacks - since you can set the value to false.

One approach I can think of:

Searching the slice can be avoided by processing and converting it to an internal map during construction of the webhook/controller. Validation can be done at that time as well (if need).


```go
// composite key that let's you flatten callbacks
type entry struct{
  gvk schema.GroupVersionKind
  op admissionv1beta1.Operation
}

type CallbackFunc func(ctx context.Context, *unstructured.Unstrucuted)error
...
// internal map
callbacks := map[entry]CallbackFunc
...
// invocation - minus error checks
cb := callbacks[entry{gvk, op}]
cb(ctx, unstructured)
```

thoughts?

Reply:
I'd just go with ctor
`NewCallback(f func(...)..., v ...webhook.Operation) *Callback`
And actually probably kept both private and and added `Do(unstructured)` method on the type.

Reply:
I went with the second suggestion which seems simplest, but perhaps that's only because it's closer to what I already had (and allows the user to pass a list and we turn it into the Set internally only).

Let me know what you think